### PR TITLE
Refactor presence update 2

### PR DESCRIPTION
**Pull request by damencho**

### CODE REVIEW

**Comment by jallamsetty1** [comment]
LGTM !!

File: JitsiConference.js [hunk 1]
```diff
@@ -1348,21 +1350,21 @@ JitsiConference.prototype._setupNewTrack = function(newTrack) {
 
 /**
  * Sets the video type.
- * @param track
+ * @param trackVideoType
  * @return <tt>true</tt> if video type was changed in presence.
  * @private
  */
-JitsiConference.prototype._setNewVideoType = function(track) {
+JitsiConference.prototype._setNewVideoType = function(trackVideoType) {
```
Comment (paweldomas) [inline JitsiConference.js:1357]: can you keep the track argument here? with multi stream support the track will be needed here

File: JitsiConference.js [hunk 2]
```diff
@@ -1371,28 +1373,28 @@ JitsiConference.prototype._setNewVideoType = function(track) {
...
 /**
  * Sets mute status.
- * @param localTrack
+ * @param mediaType
  * @param isMuted
  * @param <tt>true</tt> when presence was changed, <tt>false</tt> otherwise.
  * @private
  */
-JitsiConference.prototype._setTrackMuteStatus = function(localTrack, isMuted) {
+JitsiConference.prototype._setTrackMuteStatus = function(mediaType, isMuted) {
```
Comment (paweldomas) [inline JitsiConference.js:1392]: can you keep the local track argument here? it will be needed for multi stream support

File: JitsiConference.js [hunk 3]
```diff
@@ -3596,6 +3598,44 @@ JitsiConference.prototype._stopP2PSession = function(options = {}) {
...
 
+/**
+ * Updates room presence if needed and send the packet in case of a modification.
+ * @param {JingleSessionPC} jingleSession the session firing the event, contains the peer connection which
+ * tracks we will check.
+ * @param {Object|null} ctx a context object we can distinguish multiple calls of the same pass of updating tracks.
+ */
+JitsiConference.prototype._updateRoomPresence = function(jingleSession, ctx) {
+    if (ctx && ctx.skip) {
+        ctx.skip = true;
```
Comment (paweldomas) [inline JitsiConference.js:3609]: is it not possible to skip the call in the place which sets the `ctx`?

Comment (paweldomas) [reply JitsiConference.js:3609]: actually I don't see it being set anywhere?

Comment (damencho) [reply JitsiConference.js:3609]: https://github.com/jitsi/lib-jitsi-meet/pull/1814/files#diff-2ab65ee8ec5b7568335948135d9d2a5b7010c054a1d821f9774a7b542689b282R2525

Comment (damencho) [reply JitsiConference.js:3609]: It is just the case if you have a source add and source remove in one call of update ssrc to send one presence, not two.

File: JitsiConference.js (hunk 2 above)
Comment (damencho) [reply JitsiConference.js]: I added it back, but it is unused for now.

There is a case where you set it back to default, there is no track.

File: JitsiConference.js (hunk 1 above)
Comment (damencho) [reply JitsiConference.js]: done, but the case with going from screen share back to video muted is handled only for presence, the case when there is no video track.

File: JitsiConference.js (hunk 3 above)
Comment (paweldomas) [reply JitsiConference.js:3614]: but there it only defines the `ctx` object, where is the `ctx.skip` being set?

File: modules/xmpp/JingleSessionPC.js [hunk 4]
```diff
@@ -954,7 +954,15 @@ export default class JingleSessionPC extends JingleSession {
                 // FIXME we may not care about RESULT packet for session-accept
                 // then we should either call 'success' here immediately or
                 // modify sendSessionAccept method to do that
-                this.sendSessionAccept(success, failure);
+                this.sendSessionAccept(() => {
+                    success();
+
+                    this.room.eventEmitter.emit(XMPPEvents.SESSION_ACCEPT, this);
+                }, () => {
+                    failure();
```
Comment (paweldomas) [inline modules/xmpp/JingleSessionPC.js:962]: how is the error handled here?

File: modules/xmpp/JingleSessionPC.js [hunk 5]
```diff
@@ -2512,12 +2520,24 @@ export default class JingleSessionPC extends JingleSession {
             );
         const removedAnySSRCs = sdpDiffer.toJingle(remove);
 
+        // context a common object for one run of ssrc update (source-add and source-remove) so we can match them if we
+        // need to
+        const ctx = {};
```
Comment (paweldomas) [inline modules/xmpp/JingleSessionPC.js:2525]: what is the exact problem that `ctx` solves? It's hard to follow and passing context objects is an unusual pattern. I'm wondering if there's another of fixing that?

File: modules/xmpp/JingleSessionPC.js [hunk 6]
```diff
@@ -2538,7 +2558,16 @@ export default class JingleSessionPC extends JingleSession {
             logger.info(`${this} Sending source-add`);
             logger.debug(add.tree());
             this.connection.sendIQ(
-                add, null, this.newJingleErrorHandler(add), IQ_TIMEOUT);
+                add,
+                () => {
+                    this.room.eventEmitter.emit(XMPPEvents.SOURCE_ADD, this, ctx);
+                },
+                () => {
+                    this.newJingleErrorHandler(add);
```
Comment (paweldomas) [inline modules/xmpp/JingleSessionPC.js:2566]: newJingleErrorHandler creates a method which needs to be called with an error

File: modules/xmpp/JingleSessionPC.js [hunk 7]
```diff
@@ -2512,12 +2520,24 @@ export default class JingleSessionPC extends JingleSession {
...
             logger.debug(remove.tree());
             this.connection.sendIQ(
-                remove, null,
-                this.newJingleErrorHandler(remove), IQ_TIMEOUT);
+                remove,
+                () => {
+                    this.room.eventEmitter.emit(XMPPEvents.SOURCE_REMOVE, this, ctx);
+                },
+                () => {
+                    this.newJingleErrorHandler(remove);
```
Comment (paweldomas) [inline modules/xmpp/JingleSessionPC.js:2536]: newJingleErrorHandler creates a method which needs to be called with an error

File: JitsiConference.js (hunk 3 above)
Comment (damencho) [reply JitsiConference.js:3614]: It is set here, once it passes this method and will skip the second invocation.

Comment (damencho) [reply JitsiConference.js:3614]: Oh, I need to set it after the if 🤦‍♂️

File: modules/xmpp/JingleSessionPC.js (hunk 5 above)
Comment (damencho) [reply modules/xmpp/JingleSessionPC.js:2525]: If you do source add and source remove in one pass to send just one presence

File: JitsiConference.js (hunk 3 above)
Comment (damencho) [reply JitsiConference.js:3618]: updated.

File: modules/xmpp/JingleSessionPC.js (hunk 4 above)
Comment (damencho) [reply modules/xmpp/JingleSessionPC.js:962]: I don't know, this is the callback that was executed before.

Comment (damencho) [reply modules/xmpp/JingleSessionPC.js]: I had added the error to the callback.

File: modules/xmpp/JingleSessionPC.js (hunk 7 above)
Comment (damencho) [reply modules/xmpp/JingleSessionPC.js]: Fixed.

File: modules/xmpp/JingleSessionPC.js (hunk 6 above)
Comment (damencho) [reply modules/xmpp/JingleSessionPC.js]: Fixed.

File: JitsiConference.js (hunk 3 above)
Comment (paweldomas) [reply JitsiConference.js:3618]: I'm not a big fan of this solution, but should do for now... source-remove/source-add should be merged into a single notification.

**Review by paweldomas — APPROVED** [review]

**Comment by damencho** [comment]
Unrelated known failure, merging.
